### PR TITLE
Enable continuous controlled restore point recovery

### PR DESCRIPTION
**Pull request by soumyadeep2007**
Enable continuous controlled restore point recovery
    
This commit adds the capability to perform continuous (live) controlled
restore point based recovery for GP standbys, especially in an archive
recovery setting. This enables an application to gate recovery at
restore points created with `gp_create_restore_point()`. By gating
recovery at these restore point, eventual cluster consistency can be
achieved. A state of cluster consistency is reached when all of the
standbys in the cluster have replayed to the same restore point.

This commit introduces the `gp_pause_on_restore_point_replay` GUC, which
when set, pauses WAL recovery when a restore point WAL record is
replayed on a standby.

Example (very similar to the TAP test in the commit):

Cluster Setup:
3 primaries replicating to 3 standbys asynchronously via archive based
replication (including the coordinator and standby coordinator).

Set `gp_pause_on_restore_point_replay` ONLY on the standbys. (The GUC
should either be set before startup or set in a session followed by a
`SIGHUP` to the postmaster or a `select pg_reload_conf()` call)

1. Take a restore point on the primaries:
```sql
select gp_create_restore_point('rp');
 gp_create_restore_point
-------------------------
 (-1,0/C00026A)
 (0,0/C000260)
 (1,0/C000260)
 (2,0/C000270)
(4 rows)
```
2. Resume WAL replay on the standbys with `pg_wal_replay_resume()` on each
standby, if replication is paused. (Replication will not be paused the
first time through)

3. Force the restore point to be archived and consequently replayed on
the standbys sooner: `select gp_switch_wal()`. (this step is optional and
is purely meant for faster feedback)

4. Wait until each standby has replayed up to its corresponding
restore point LSN (match by content ID). Each standby will be paused at
that LSN (can be verified with `pg_current_replay_lsn()`).
```
-1 will be paused at 0/C00026A
0  will be paused at 0/C000260
1  will be paused at 0/C000260
2  will be paused at 0/C000270
```
5. Go to 1.

Notes:

* The recovery_target_XXX GUCs cannot be used to realize this feature as
they are only applicable to PITR. This means that once the recovery
target is reached, subsequent updates to these GUCs are ignored in
standby mode.

### CODE REVIEW

File: src/backend/access/transam/xlog.c [hunk 1]
```diff
@@ -5992,6 +5999,14 @@ SetRecoveryPause(bool recoveryPause)
 	SpinLockRelease(&XLogCtl->info_lck);
 }
 
+void
+SetRecoveryPauseRestorePoint(XLogRecPtr restorePointLocation)
+{
+	SpinLockAcquire(&XLogCtl->info_lck);
+	XLogCtl->recoveryPauseRestorePoint = restorePointLocation;
```
Comment (asimrp) [inline src/backend/access/transam/xlog.c:6006]: I'm wondering why can't the `recoveryStopsBefore` or `recoveryStopsAfter` facility in upstream be used for this purpose, instead of having to introduce a new member in XLogCtl shared struct?

File: src/test/recovery/t/101_restore_point_and_startup_pause.pl [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+# Check if recovery is paused at startup
+my $result1 = $node_standby->safe_psql('postgres', "SELECT pg_is_wal_replay_paused()");
+is($result1, qq(t), 'check if WAL replay is paused at startup');
+
+# Resume WAL replay after startup pause.
+$node_standby->safe_psql('postgres', "SELECT pg_wal_replay_resume()");
+
+# Create a restore point
+my $restore_point_lsn =
+    $node_primary->safe_psql('postgres', "SELECT pg_create_restore_point('rp')");
```
Comment (asimrp) [inline src/test/recovery/t/101_restore_point_and_startup_pause.pl:40]: Ah, we are already using the upstream facility to some extent.  It is still not clear to me where the upstream functionality falls short...

File: src/backend/access/transam/xlog.c (hunk 1 above)
Comment (soumyadeep2007) [reply src/backend/access/transam/xlog.c:6006]: `recoveryStopsXXX` class of functions are only called in a PITR context.
To expand on what I mentioned in my commit message as to why we can't use the `recovery_target_XXX` GUCs for this feature:
The GUC values are only checked/useful when PITR is being performed. 

1. In `startupXLOG()`, whenever we reach the "point" specified by the `recovery_target_XXX` GUCs, `reachedStopPoint` is set.
2. This leads to a break from the main redo loop and recovery is over.
3. Then based on the GUC recovery_target_action we either `shutdown`, `pause` or `promote`. For our purposes, we would set it as `pause`
See `switch (recoveryTargetAction)`
4. Recovery ends and we switch to a new timeline.
See `ThisTimeLineID = findNewestTimeLine(recoveryTargetTLI) + 1;`

What we want is continuous recovery. We don't want recovery to end as soon as the recovery target is reached. We want to keep pausing and resuming recovery from one restore point to the next. Since recovery ends immediately on reaching the recovery target, even if the action is pause, a subsequent `pg_wal_replay_resume()` eventually ends recovery (See `recoveryPausesHere()`)

So, in conclusion, `recovery_target_XXX` GUCs and the `recoveryStopsXXX` functions are meant only for PITR.

File: src/backend/access/transam/xlog.c [hunk 3]
```diff
@@ -7193,8 +7208,21 @@ StartupXLOG(void)
...
 
+		/*
+		 * GPDB: Pause recovery initially if GUC is set so that the
+		 * recovery process doesn't just start replaying WAL asap.
+		 */
+		if (gp_wal_pause_on_recovery_start)
+		{
+			SetRecoveryPause(true);
+			ereport(LOG,
+					(errmsg("recovery has paused as gp_wal_pause_on_recovery_start is set"),
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:7222]: well recovery is not logically paused here only set to be paused later in flow. So, this read wrong and will be confusing when emitted in logs as well when someone traces back to code.

File: src/backend/access/transam/xlog.c [hunk 4]
```diff
@@ -10270,7 +10298,15 @@ xlog_redo(XLogReaderState *record)
 	}
 	else if (info == XLOG_RESTORE_POINT)
 	{
-		/* nothing to do here */
+		xl_restore_point xlrec;
+		bool			 pause;
+
+		memcpy(&xlrec, XLogRecGetData(record), sizeof(xl_restore_point));
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:10304]: I am confused, what role `xlrec` plays here? Seems we are not using it for anything.

File: src/backend/access/transam/xlogfuncs.c [hunk 5]
```diff
@@ -565,6 +565,28 @@ pg_wal_replay_pause(PG_FUNCTION_ARGS)
 	PG_RETURN_VOID();
 }
 
+Datum
+pg_wal_replay_pause_on_restore_point(PG_FUNCTION_ARGS)
```
Comment (ashwinstar) [inline src/backend/access/transam/xlogfuncs.c:569]: why `pg` in name, its `gp` specific function, right?

**Review by ashwinstar — COMMENTED** [review]
I was not able to give full review to this PR today. Still would like to think on the flow and such. Reads confusing to have Greenplum specific logic mixed with some of upstream code and variables. Will try to get to it tomorrow.

File: src/backend/access/transam/xlog.c [hunk 6]
```diff
@@ -118,6 +118,10 @@ int			CommitDelay = 0;	/* precommit delay in microseconds */
 int			CommitSiblings = 5; /* # concurrent xacts needed to sleep */
 int			wal_retrieve_retry_interval = 5000;
 
+/* GPDB specific */
+bool		gp_wal_pause_on_recovery_start = false;
```
[the remark this answers is not in the document]
Comment (jimmyyih) [reply src/backend/access/transam/xlog.c:122]: What about `gp_pause_replay_on_recovery_start`?  The GUC specifically only does an initial pause when recovery has started (before the main recovery loop) and doesn't control anything else during recovery.

File: src/backend/access/transam/xlog.c (hunk 4 above)
Comment (soumyadeep2007) [reply src/backend/access/transam/xlog.c:10304]: Whoops forgot to remove that. Will do, thanks.

File: src/backend/access/transam/xlogfuncs.c (hunk 5 above)
Comment (soumyadeep2007) [reply src/backend/access/transam/xlogfuncs.c:569]: True. I left the name as `pg` because it operates on a single PG server. I will change it to `gp`

File: src/backend/access/transam/xlog.c [hunk 7]
```diff
@@ -730,6 +734,9 @@ typedef struct XLogCtlData
 	XLogRecPtr	lastFpwDisableRecPtr;
 
 	slock_t		info_lck;		/* locks shared variables shown above */
+
+	/* LSN of the next restore point to pause recovery on */
+	XLogRecPtr	recoveryPauseRestorePoint;
```
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/backend/access/transam/xlog.c:739]: Hmm, I don't agree. It is the LSN of a specific restore point that we want to pause on.

[the remark this answers is not in the document]
Comment (ashwinstar) [reply src/backend/access/transam/xlog.c:739]: well that's the main aspect I wish to bring up, what does "restore point" mean in context on continuous pause and replay. It's actually "consistency point" I guess. Restore point sounds more PITR notion to restore to database back to some point.

[the remark this answers is not in the document]
Comment (jimmyyih) [reply src/backend/access/transam/xlog.c:739]: Agree with Ashwin here that `recoveryPauseLSN` makes more sense.  The UDF has no validation that the provided LSN is a restore point LSN (not possible either way).  It simply just pauses on an LSN so its required variable should also be generically named.

[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/backend/access/transam/xlog.c:739]: I wanted to make clear that we pause only on a restore point's LSN (We only compare the LSN to the requested pause LSN when we [replay a restore point record](https://github.com/greenplum-db/gpdb/pull/11621/files#diff-c1cb3ab2a19606390c1a7ed00ffe5a45531702ca5faf999d401c548f8951c65bR10306)).

> It simply just pauses on an LSN so its required variable should also be generically named.
FYI the UDF doesn't immediately pause as Ashwin pointed out. It requests a pause.

Now, question is, should we expose general functionality to pause on any pre-specified LSN, that kicks in when the LSN is encountered during replay?

One may argue that restore point based pauses might be a very specific use case (we overload restore points as cluster consistency points)

Is that what you are talking about @ashwinstar ?

[the remark this answers is not in the document]
Comment (jimmyyih) [reply src/backend/access/transam/xlog.c:739]: Okay sorry, I forgot how things kinda worked for a second.  I would say make it `recoveryPauseRestorePointLSN`.  The variable is indeed directly associated with restore points as the variable is only used for LSN comparisons when replaying restore point records.

In that case, should we be more descriptive in the new catalog function as well and name it `gp_recovery_pause_on_restore_point_lsn()`?

> FYI the UDF doesn't immediately pause as Ashwin pointed out. It requests a pause.

True that it doesn't immediately pause and only requests but that request is immediately invoked after the restore point WAL record is handled.  It's essentially as good as saying it pauses recovery on the user-provided lsn.

**Comment by ashwinstar** [comment]
It will help me as reviewer (tomorrow) and for long term if comment can be added with clear flow of events on how newly added GUCs and functions work taking some hypothetical example LSNs maybe.

**Comment by soumyadeep2007** [comment]
@ashwinstar @jimmyyih Ready for another round of review. I have included the workflow in the commit message, which is very similar to the TAP test in the commit.

File: src/backend/utils/misc/guc_gp.c [hunk 8]
```diff
@@ -2784,6 +2784,18 @@ struct config_bool ConfigureNamesBool_gp[] =
 		false,
 		NULL, NULL, NULL
 	},
+
+	{
+		{"gp_pause_replay_on_recovery_start", PGC_SIGHUP, DEVELOPER_OPTIONS,
```
Comment (ashwinstar) [inline src/backend/utils/misc/guc_gp.c:2789]: So this GUC has to be always set to ON (mostly via PostgreSQL conf file) for such kind of GP Standbys, correct? As even on crash / restart wish it to pause and wait to get instructed to replay till what restore point. So, do we set this GUC to ON when creating such standby?

Other possibility could be instead of this specific GUC to have a generic GUC which conveys this is continuous pausing - replaying kind of standby something like `hot_standby` GUC. Then such GUC maybe reused for other purposes as well where functionality needs to differ from regular standby, just a thought.

File: src/backend/access/transam/xlog.c [hunk 9]
```diff
@@ -7193,8 +7208,21 @@ StartupXLOG(void)
...
+		/*
+		 * GPDB: Pause recovery initially if GUC is set so that the
+		 * recovery process doesn't just start replaying WAL asap.
+		 */
+		if (gp_pause_replay_on_recovery_start)
+		{
+			SetRecoveryPause(true);
+			ereport(LOG,
+					(errmsg("gp_pause_replay_on_recovery_start set: replay will be paused at recovery start"),
+						errhint("Execute pg_wal_replay_resume() to continue.")));
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:7223]: `errhint()` comes with danger of calling `pg_wal_replay_resume()` without setting restorepoint LSN. Then replay will run till end of xlog. Do we wish to convey about setting restore point as well?

File: src/backend/access/transam/xlog.c [hunk 10]
```diff
@@ -7193,8 +7208,21 @@ StartupXLOG(void)
...
 		XLogCtl->currentChunkStartTime = 0;
 		XLogCtl->recoveryPause = false;
+		XLogCtl->recoveryPauseRestorePointLSN = InvalidXLogRecPtr;
 		SpinLockRelease(&XLogCtl->info_lck);
 
+		/*
+		 * GPDB: Pause recovery initially if GUC is set so that the
+		 * recovery process doesn't just start replaying WAL asap.
+		 */
+		if (gp_pause_replay_on_recovery_start)
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:7218]: Any specific reason the code is located here.. seems would be better if placed just before "main redo apply loop" for better code readability.

File: src/backend/access/transam/xlog.c [hunk 11]
```diff
@@ -10270,7 +10298,13 @@ xlog_redo(XLogReaderState *record)
 	}
 	else if (info == XLOG_RESTORE_POINT)
 	{
-		/* nothing to do here */
+		bool			 pause;
```
Comment (ashwinstar) [inline src/backend/access/transam/xlog.c:10301]: Please can you write comment and explain what all is this for? Reads fishy :-)
One flow is pause, to set restore point and then this one seems if restorepoint is set, pause without even using any info in the record. Just blindly leveraging any `XLOG_RESTORE_POINT` to pause the recovery. Not comforting code.

File: src/backend/access/transam/xlogfuncs.c [hunk 12]
```diff
@@ -565,6 +565,28 @@ pg_wal_replay_pause(PG_FUNCTION_ARGS)
...
+				 errmsg("recovery is not in progress"),
+				 errhint("Recovery control functions can only be executed during recovery.")));
+
+	if (!XLogIsNeeded())
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				 errmsg("WAL level not sufficient for creating a restore point"),
+				 errhint("wal_level must be set to \"replica\" or \"logical\" at server start.")));
+
+	SetRecoveryPauseRestorePointLSN(restorePointLocation);
```
Comment (ashwinstar) [inline src/backend/access/transam/xlogfuncs.c:585]: Should we also check recovery is paused just as added check?

**Review by ashwinstar — COMMENTED** [review]
Thanks @soumyadeep2007 the mentioned flow really helped to speed up the review. From the test the usage of GUC `gp_pause_replay_on_recovery_start` is not clear as test right away resumes the execution. Then why even have the GUC as doing nothing, the comment helped. I think we need to reflect that comment in code to clarity the usage.

**Comment by ashwinstar** [comment]
Curious, similar to `select gp_create_restore_point('rpx2');` do we not have function to `select gp_replay_upto_restore_point(<name>)` instead of individual LSN values? I think using name can simplify the logic by having unique names for restorePoints (maybe using coordinators current insert LSN value or something on those lines) and then using that.

File: src/backend/access/transam/xlog.c (hunk 11 above)
Comment (leskin-in) [reply src/backend/access/transam/xlog.c:10301]: As far as I understand, this code implements the main point of this PR.

The goal is to be able to pause the recovery exactly when some restore point is reached. This leaves the GPDB cluster being recovered ("replica") in a known consistent state - the one obtained at this restore point. As a result, in case of a failure of the "original" cluster, a "replica" can replace it, and the resulting cluster state is known.

This can already be achieved using existing restore point machinery; however, currently it is only possible to pause the "replica" just once (before recovery finish). This code makes it possible to pause the "replica" multiple times. During these pauses, extra WALs could be provided. As a result, the "replica" would be continuously updated and paused always at consistent state.

**Comment by leskin-in** [comment]
I suppose the function [proposed](https://github.com/greenplum-db/gpdb/pull/11621#issuecomment-802297077) by @ashwinstar would be more convenient than the existing one in some use cases. LSNs on different GPDB segments may differ; it seems more convenient to issue the same function call across all GPDB segments under recovery than issue multiple function calls with different LSNs as their parameters.

**Review by soumyadeep2007 — COMMENTED** [review]
> Curious, similar to select gp_create_restore_point('rpx2'); do we not have function to select gp_replay_upto_restore_point(<name>) instead of individual LSN values? I think using name can simplify the logic by having unique names for restorePoints (maybe using coordinators current insert LSN value or something on those lines) and then using that.

That is what our initial design was. However there are a few reasons why we deviated from it:

1. Supplying unique names for restore points is not straightforward (and if we are going to embed an LSN in the name, might as well use the LSN directly)
2. LSNs are guaranteed to be unique, whereas names are not. Consider what would happen if there are interleaved `pg_create_restore_point()` and `gp_create_restore_point()` calls **with the same restore point name**. We would end up pausing on the restore point records created with `pg_create_restore_point()` as well - when our intention might be to only pause  on the ones created by `gp_create_restore_point()`

   Alternatively, imagine where we have taken a `gp_create_restore_point()`, for which there was a failure on one of the segments and we have to retry `gp_create_restore_point()` with the same restore point name (all of the segments will have 2 restore point records with the same name, except the failed segment which will have 1). We would want to pause on the WAL records corresponding to the second invocation of `gp_create_restore_point()`. Unfortunately, we would have paused on the records corresponding to the first one as well.
3. Restore point LSNs are easy enough to obtain from `pg_create_restore_point()` and `gp_create_restore_point()` return values.

Using an LSN thus gives much more precision in gating WAL replay. I have added a note about this to the commit message.

**Comment by soumyadeep2007** [comment]
> I suppose a function similar to the one proposed by @ashwinstar would be more convenient than the one introduced by this PR (gp_recovery_pause_on_restore_point_lsn(<restore_point_LSN>)) in some use cases. LSNs on different GPDB segments may differ; it seems more convenient to issue the same function call across all GPDB segments under recovery than issue multiple function calls with different LSNs as their parameters.

See my comment [here](https://github.com/greenplum-db/gpdb/pull/11621#pullrequestreview-615876875)

**Comment by jimmyyih** [comment]
A function that takes in a restore point name to pause on would indeed be more convenient but there would definitely be application overhead costs to do validations (like the examples that Deep had given).  IMO, we could always have both (kinda like how PITR has name, lsn, xid, and timestamp recovery targets) and this PR would implement the lsn aspect.  Whoever wants the other ones can open up a PR.. although for GPDB, I think name (convenience) and lsn (precision) would be the only useful ones for cluster consistency.

File: src/backend/access/transam/xlog.c [hunk 13]
```diff
@@ -10270,7 +10296,19 @@ xlog_redo(XLogReaderState *record)
...
+		 * preemptive pause could have been set with a call to
+		 * gp_recovery_pause_on_restore_point_lsn().
+		 */
+		bool			 pause;
+
+		SpinLockAcquire(&XLogCtl->info_lck);
+		pause = lsn == XLogCtl->recoveryPauseRestorePointLSN;
+		SpinLockRelease(&XLogCtl->info_lck);
+		if (pause)
+			SetRecoveryPause(true);
```
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/backend/access/transam/xlog.c:10311]: > Would it be simpler (less Greenplum-specific code) if recovery is set to pause after replaying every restore point record? That is, each segment on the replica site replays WAL pausing at every restart point. It needs to be explicitly unpaused. The pg_last_wal_replay_lsn() can be used to determine which restartpoint a specific segment is paused at. Additional code may be written to make this convenient, e.g. gp_last_wal_replay_lsn() to obtain replay LSNs from all segments and gp_wal_replay_resume() to resume WAL replay until the next restart point on all segments.

I think you mean restore point instead of restart point (since restartpoints are actions taken on checkpoints during recovery)?
I think what you are talking about is what our initial design was. [This](https://github.com/greenplum-db/gpdb/pull/11621#pullrequestreview-615876875) details why we moved away from that strategy (especially point 2).

**Comment by asimrp** [comment]
> Consider what would happen if there are interleaved pg_create_restore_point() and gp_create_restore_point() calls ...

It is hard to imagine a valid use case in Greenplum's distributed architecture for `pg_create_restore_point()`.  This function should be treated like utility mode connections are and `gp_create_restore_point()` should be used instead.  Does that make sense?

File: src/backend/access/transam/xlog.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (asimrp) [reply src/backend/access/transam/xlog.c:10311]: Sorry for making that mistake repeatedly, yes I mean restore point.  What's in a name, huh? :-)

**Comment by asimrp** [comment]
I am imagining this patch to be as simple as:
```diff
diff --git a/src/backend/access/transam/xlog.c b/src/backend/access/transam/xlog.c
index 1ea70ea37d1..bdb7ccd6fe0 100644
--- a/src/backend/access/transam/xlog.c
+++ b/src/backend/access/transam/xlog.c
@@ -10270,7 +10270,13 @@ xlog_redo(XLogReaderState *record)
        }
        else if (info == XLOG_RESTORE_POINT)
        {
-               /* nothing to do here */
+               /*
+                * Greenplum: restore point record is repurposed as a point of
+                * synchronization to ensure cluster-wide consistency during WAL
+                * replay.  WAL replay is paused at each restore point until it is
+                * explicitly unpaused (by the coordinator?).
+                */
+               SetRecoveryPause(true);
        }
        else if (info == XLOG_FPI || info == XLOG_FPI_FOR_HINT)
        {                                                                                                                                                                                                                                      
```
and then a new convenience function (after making dispatch work from hot standby mode) `gp_wal_replay_resume()` that can resume WAL replay on all segments such that they replay till the next restore point.  With this approach, it seems that the proposed GUC may be eliminated too?

The underlying assumption is creating restore points directly on segments using `pg_create_restore_point()` is disallowed.  If that is not possible, would it be better to invent a new Greenplum-specific XLOG record, `XLOG_GP_SYNC_POINT`, and then create Greenplum-specific UDFs to expose its creation?

**Comment by leskin-in** [comment]
@asimrp , one implication of your [most recent comment](https://github.com/greenplum-db/gpdb/pull/11621#issuecomment-811840851) is that the _recovery without stops at restore points becomes impossible_. As far as I understand, this may break replication to (usual) GPDB mirrors, because they use the same recovery process to replay incoming WAL records.

**Comment by soumyadeep2007** [comment]
I agree with the comment from @leskin-in above.

> It is hard to imagine a valid use case in Greenplum's distributed architecture for pg_create_restore_point(). This function should be treated like utility mode connections are and gp_create_restore_point() should be used instead. Does that make sense?

This is definitely an edge case. But the case where a gp_create_restore_point()
call may fail on certain servers and succeed on others is more probable.

> and then a new convenience function (after making dispatch work from hot standby mode) gp_wal_replay_resume() that can resume WAL replay on all segments such that they replay till the next restore point. With this approach, it seems that the proposed GUC may be eliminated too?

Well, we would still need the GUC as we would have to pause replay at the
beginning of startup recovery.

> The underlying assumption is creating restore points directly on segments using pg_create_restore_point() is disallowed. If that is not possible, would it be better to invent a new Greenplum-specific XLOG record, XLOG_GP_SYNC_POINT, and then create Greenplum-specific UDFs to expose its creation?

Idk if we want to disallow this explicitly - one use case that can be lost is if
someone wants to PITR each segment separately with restore points, they will not
be able to do so (although this use case is contrived). We should not lose the
fact that PG instances in GP can be operated on piecemeal.

Currently, the only way to create a restore point record is to use
`pg_create_restore_point()`, which means we will have to expose creation of
restore points in a different manner, were we to ban it.

`XLOG_GP_SYNC_POINT` does not get us anything extra since we still would
have to deal with partially successful sync point creations. Ultimately, we would
end up with duplicating code from restore points and we would still need the
preemptive pausing method introduced here.

-
Deep and Jimmy

**Comment by soumyadeep2007** [comment]
After an offline discussion between @asimrp , @jimmyyih  and myself, we arrived at the consensus to revert to a slightly modified version of our original solution: 

* Control recovery through a single GUC: `gp_pause_on_restore_point_replay` which will pause replay at **every** restore point encountered in the WAL stream.
* Don't pause recovery at startup: all the standbys can pause at the earliest restore point after normal startup and at the first restore point that follows in the WAL stream during crash recovery.
* Eliminate the LSN targeted preemptive pause functionality that we had.

The result is a reduction in complexity for the server and reduced divergence with upstream. This comes at the cost of pushing the complexity of pausing/resuming when partially-failed or rogue restore restore points are present in the WAL stream, to the external program that will control cluster-wide WAL replay.

The code and commit message have been updated. Please take a look @asimrp.

**Review by asimrp — APPROVED** [review]
Thank you for patiently explaining things thus far, the PR is looking great now.

I’m wondering if the catalog version bump is necessary.  Seems ok to keep it though.

**Comment by soumyadeep2007** [comment]
> Thank you for patiently explaining things thus far, the PR is looking great now.
Our pleasure! Thanks for the thorough review!

> I’m wondering if the catalog version bump is necessary. Seems ok to keep it though.
Nice catch. It is indeed unnecessary, changes in gucs are not accompanied by catalog changes.
